pennylane/workflow/run.py

```diff
@@ -0,0 +1,317 @@
+                return device.compute_derivatives(numpy_tapes, config)
+
+            diff_method = device_compute_derivatives
+
+    elif config.grad_on_execution is True:
+        raise ValueError("Gradient transforms cannot be used with grad_on_execution=True")
+
+    else:
+        diff_method = config.gradient_method
+
+    return execute_fn, diff_method
+
+
+def _construct_ml_execution_pipeline(config, device, inner_transform_program):
+    """Constructs the ML execution pipeline for all interfaces except TF_AUTOGRAPH.
+
+    This function determines the execution function (`execute_fn`) and the Jacobian product
+    class (`jpc`) required for gradient computations.
+
+    Args:
+        config (qml.devices.ExecutionConfig): resolved execution configuration
+        device (qml.devices.Device): a Pennylane device
+        inner_transform_program (qml.transforms.core.TransformProgram): the transformation applied to quantum tapes
+
+    Returns:
+        tuple: A tuple containing:
+            - `jpc`: jacobian product class for computing gradients
+            - `execute_fn`: the function to execute quantum tapes
+
+    Raises:
+        ValueError: If gradients are computed on execution (`grad_on_execution=True`).
+    """
+    inner_execute = _make_inner_execute(device, inner_transform_program, config)
+    cache = _cache_transform in inner_transform_program
+
+    execute_fn = inner_execute
+    jpc = None
+
+    if config.use_device_jacobian_product:
+        jpc = DeviceJacobianProducts(device, config)
+
+    elif config.use_device_gradient:
+        jpc = DeviceDerivatives(device, config)
+        if config.grad_on_execution:
+            execute_fn = jpc.execute_and_cache_jacobian
+        else:
+            execute_fn = inner_execute
+
+    elif config.grad_on_execution is True:
+        raise ValueError("Gradient transforms cannot be used with grad_on_execution=True")
+
+    else:
+        cache_full_jacobian = (config.interface == Interface.AUTOGRAD) and not cache
+        jpc = TransformJacobianProducts(
+            execute_fn,
+            config.gradient_method,
+            config.gradient_keyword_arguments,
+            cache_full_jacobian,
+        )
+        for i in range(1, config.derivative_order):
+            differentiable = i > 1
+            ml_boundary_execute = _get_ml_boundary_execute(config, differentiable=differentiable)
+            execute_fn = partial(
+                ml_boundary_execute,
+                execute_fn=execute_fn,
+                jpc=jpc,
+                device=device,
+            )
+            jpc = TransformJacobianProducts(
+                execute_fn,
+                config.gradient_method,
+                config.gradient_keyword_arguments,
+            )
+
+    return jpc, execute_fn
+
+
+# pylint: disable=import-outside-toplevel
+def _get_ml_boundary_execute(
+    resolved_execution_config: "qml.devices.ExecutionConfig", differentiable=False
```

Comment:
Do we still need to use the strings, or can we now import this here and have `qml.devices.ExecutionConfig`?

Reply:
Unfortunately, we still need strings 😢 . Lots of circular import issues that aren't trivial to sort out.